### PR TITLE
TikTak Optimization

### PR DESCRIPTION
**Pull request by asouther4**
See [this paper](https://www.nber.org/system/files/working_papers/w26340/w26340.pdf) for a description of the TikTak algorithm. Alisdair McKay has [implemented](https://github.com/amckay/TikTak) the algorithm in Python already

I have edited McKay's code and started integrating TikTak into estimagic. Still a work in progress, but I'm opening up the pull request so further changes will be public. I've also pushed a testing file to the directory `tests\optimization`.

Things I can work on soon: 
- ~~allowing the user to provide a set of custom starting points~~
- ~~editing the "mixing weight" formula to more closely match the formula described on page 46 of the paper above.~~
- Building a more thorough testing file

Things I'll probably need help with:
- logging
- parallelization
- error handling

Edit: striking through finished tasks

### CODE REVIEW

**Comment by janosg** [comment]
Hi Andrew, thanks a lot for this PR and sorry for the delay! Your code helps me tremendously to understand the TikTak algorithm. I think I can take it from here, address the comments you mentioned and do the integration into estimagic's `minimize` and `maximize` functions!

**Comment by asouther4** [comment]
> Hi Andrew, thanks a lot for this PR and sorry for the delay! Your code helps me tremendously to understand the TikTak algorithm. I think I can take it from here, address the comments you mentioned and do the integration into estimagic's `minimize` and `maximize` functions!

Good to hear, thanks for picking this up, Janos! Let me know if there are any features you'd like me to work on.

File: src/estimagic/optimization/tiktak.py [hunk 1]
```diff
@@ -0,0 +1,569 @@
...
+    np.random.seed(seed)
+
+    sample = joint_distribution.sample(
+        size=n_samples,
+        rule=sampling_method,
+    ).T
+    return sample
+
+
+def get_internal_sampling_bounds(params, constraints):
```
Comment (roecla) [inline src/estimagic/optimization/tiktak.py:290]: Pass in processed constraints.

**Review by roecla — APPROVED** [review]
:tada:

File: src/estimagic/optimization/optimize.py [hunk 2]
```diff
@@ -604,71 +607,75 @@ def _single_optimize(
...
-        "n_iterations",
-        "success",
-        "reached_convergence_criterion",
-        "message",
-    ]
+        multistart_options = _fill_multistart_options_with_defaults(
+            multistart_options,
+            params,
+            constraints,
+        )
```
[the remark this answers is not in the document]
Comment (janosg) [reply src/estimagic/optimization/optimize.py:655]: Done.

File: src/estimagic/optimization/tiktak.py [hunk 3]
```diff
@@ -0,0 +1,569 @@
...
+    x,
+    lower_bounds,
+    upper_bounds,
+    options,
+    logging,
+    db_kwargs,
+    error_handling,
+    error_penalty,
+):
+    steps = determine_steps(options["n_samples"], options["n_optimizations"])
```
[the remark this answers is not in the document]
Comment (janosg) [reply src/estimagic/optimization/tiktak.py:48]: Done.

**Pull request merged**